### PR TITLE
GH-113486: Do not emit spurious `PY_UNWIND` events for optimized calls to classes.

### PR DESCRIPTION
**Pull request by markshannon**
Adds the `CO_NO_MONITORING_EVENTS` flag for use by shim frame code objects and similar, to avoid monitoring events.

* Issue: gh-113486

### CODE REVIEW

File: Misc/NEWS.d/next/Core and Builtins/2023-12-31-07-46-01.gh-issue-113486.uki19C.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+No longer issue spurious PY_UNWIND events for optimized calls to classes
```
Comment (gvanrossum) [inline Misc/NEWS.d/next/Core and Builtins/2023-12-31-07-46-01.gh-issue-113486.uki19C.rst]:
```suggestion
No longer issue spurious ``PY_UNWIND`` events for optimized calls to classes.
```

File: Python/specialize.c [hunk 2]
```diff
@@ -2534,7 +2534,7 @@ const struct _PyCode_DEF(8) _Py_InitCleanup = {
     .co_consts = (PyObject *)&_Py_SINGLETON(tuple_empty),
     .co_names = (PyObject *)&_Py_SINGLETON(tuple_empty),
     .co_exceptiontable = (PyObject *)&_Py_SINGLETON(bytes_empty),
-    .co_flags = CO_OPTIMIZED,
+    .co_flags = CO_OPTIMIZED | CO_NO_MONITORING_EVENTS,
     .co_localsplusnames = (PyObject *)&_Py_SINGLETON(tuple_empty),
     .co_localspluskinds = (PyObject *)&_Py_SINGLETON(bytes_empty),
     .co_filename = &_Py_ID(__init__),
```
Comment (gvanrossum) [inline Python/specialize.c:2540]: Do we still need to set `co_firsttraceable` to 4 below?

**Review by gvanrossum — APPROVED** [review]

File: Python/specialize.c (hunk 2 above)
Comment (markshannon) [reply Python/specialize.c:2540]: It has to have some value, and anything less would be misleading.

**Comment by nedbat** [comment]
Thanks for the quick action.  I took the liberty of testing this branch.  The events reported to the settrace function is changed:
```py
import linecache, sys

def trace(frame, event, arg):
    # The weird globals here is to avoid a NameError on shutdown...
    if frame.f_code.co_filename == globals().get("__file__"):
        lineno = frame.f_lineno
        line = linecache.getline(__file__, lineno).rstrip()
        print("{} {}: {}".format(event[:4], lineno, line))
    return trace

print(sys.version)
sys.settrace(trace)

def gen():
    print(2)
    yield 3
    print(4)

print(next(gen()))
```
Running with 3.12:
```
3.12.1 (main, Dec  8 2023, 20:58:19) [Clang 15.0.0 (clang-1500.0.40.1)]
call 14: def gen():
line 15:     print(2)
2
line 16:     yield 3
retu 16:     yield 3
call 16:     yield 3
exce 16:     yield 3
retu 16:     yield 3
3
```
Running with this branch:
```
3.13.0a2+ (heads/pr/113680:e92fd7b816, Jan  4 2024, 10:22:55) [Clang 15.0.0 (clang-1500.1.0.2.5)]
call 14: def gen():
line 15:     print(2)
2
line 16:     yield 3
retu 16:     yield 3
3
```

Do you mean to skip the return and call events from the yield?

**Comment by nedbat** [comment]
BTW, the 3.13 behavior also happens with 3.13.0a2, so it's not due to the change in this branch, but seems related?
```
3.13.0a2 (main, Nov 24 2023, 14:53:47) [Clang 15.0.0 (clang-1500.0.40.1)]
call 14: def gen():
line 15:     print(2)
2
line 16:     yield 3
retu 16:     yield 3
3
```

**Comment by markshannon** [comment]
@nedbat 
I've opened https://github.com/python/cpython/issues/113728 for the difference between 3.12 and 3.13.

**Comment by gaogaotiantian** [comment]
Sorry I was traveling and did not have a reliable network access. But the change looks good to me.

**Pull request merged**